### DISCUSSION TITLE
Structural refactor

### DISCUSSION THREAD

**Discussion by kartik4949** [xinntao/Real-ESRGAN#75] [Ideas]
Hi,
I can help adding losses zoo, user can pick a loss stated in a config file

**Comment by kartik4949** [comment]
lets also add drivers i.e trainer, inference driver, input data validator

**Comment by kartik4949** [comment]
we can levearge my input pipline library tensorpipe, which is a high performance library with all state of the art augmenations for computer vision.

**Comment by xinntao** [comment]
Thanks for the discussion 😄 

1. The overall structures, such as the loss zoo, trainer, etc, are designed to be provided by BasicSR: https://github.com/xinntao/BasicSR
2. I use BasicSR as the overall framework in Real-ESRGAN. And in this repo, I hope to focus on the practical image restoration itself, such as improving the performance for different categories of inputs, or speeding up the inference time.  
3. For your suggestions, I think we can discuss whether it is possible to have something improved in BasicSR, which servers as a handy training framework.

**Reply by kartik4949** [reply]
Hi, 
Got it, i saw you are importing basicSR utilities for train, infere and all!
I guess we can add inference, loss zoo and all in basicSR then 
:)

**Reply by kartik4949** [reply]
Also please checkout my work 
https://www.github.com/kartik4949/TensorPipe

https://www.github.com/kartik4949/AutoDeploy
